pkg/controller/replica.go

```diff
@@ -0,0 +1,335 @@
+}
+func (r *ReconcileReplica) getClaimToVolumeMap() *SharedMap {
+	return r.claimToVolumeMap
+}
+func (r *ReconcileReplica) getVolumeToClaimMap() *SharedMap {
+	return r.volumeToClaimMap
+}
+func (r *ReconcileReplica) getClient() client.Client {
+	return r.client
+}
+func (r *ReconcileReplica) getAzClient() azClientSet.Interface {
+	return r.azVolumeClient
+}
+func (r *ReconcileReplica) getNamespace() string {
+	return r.namespace
+}
+
+var _ reconcile.Reconciler = &ReconcileReplica{}
+
+func (r *ReconcileReplica) Reconcile(ctx context.Context, request reconcile.Request) (reconcile.Result, error) {
+	azVolumeAttachment, err := azureutils.GetAzVolumeAttachment(ctx, r.client, r.azVolumeClient, request.Name, request.Namespace, true)
+	if errors.IsNotFound(err) {
+		klog.Infof("AzVolumeAttachment (%s) has been successfully deleted.", request.Name)
+		return reconcile.Result{}, nil
+	} else if err != nil {
+		klog.Errorf("failed to fetch AzVolumeAttachment (%s): %v", request.Name, err)
+		return reconcile.Result{Requeue: true}, err
+	}
+
+	if azVolumeAttachment.Spec.RequestedRole == v1alpha1.PrimaryRole {
+		// Deletion Event
+		if requestedToBeDeleted(azVolumeAttachment) {
+			// If primary attachment is marked for deletion, queue garbage collection for replica attachments
+			r.triggerGarbageCollection(azVolumeAttachment.Spec.UnderlyingVolume)
+		} else {
+			// If not, cancel scheduled garbage collection if there is one enqueued
+			r.cancelGarbageCollection(azVolumeAttachment.Spec.UnderlyingVolume)
+
+			// If promotion event, create a replacement replica
+			if isAttached(azVolumeAttachment) && azVolumeAttachment.Status.Detail.Role != azVolumeAttachment.Spec.RequestedRole {
+				if err := r.manageReplicas(ctx, azVolumeAttachment.Spec.UnderlyingVolume); err != nil {
+					return reconcile.Result{Requeue: true}, err
+				}
+			}
+		}
+	} else {
+		// create a replacement replica if replica attachment failed or promoted
+		if requestedToBeDeleted(azVolumeAttachment) {
+			if azVolumeAttachment.Annotations == nil || !metav1.HasAnnotation(azVolumeAttachment.ObjectMeta, azureutils.CleanUpAnnotation) {
+				go func() {
+					conditionFunc := func() (bool, error) {
+						var tmp v1alpha1.AzVolumeAttachment
+						err = r.client.Get(ctx, request.NamespacedName, &tmp)
+						return errors.IsNotFound(err), nil
+					}
+					_ = wait.PollImmediateInfinite(azureutils.DefaultAttachTimeout/2, conditionFunc)
+
+					for i := 0; i < maxRetry; i++ {
+						if err := r.manageReplicas(ctx, azVolumeAttachment.Spec.UnderlyingVolume); err == nil {
+							break
+						}
+					}
+				}()
+			}
+		}
+	}
+	return reconcile.Result{}, nil
+}
+
+func (r *ReconcileReplica) triggerGarbageCollection(volumeName string) {
+	emptyCtx := context.TODO()
+	deletionCtx, cancelFunc := context.WithCancel(emptyCtx)
+	r.cleanUpMap.add(volumeName, cancelFunc)
+	klog.Infof("garbage collection for AzVolume (%s) scheduled in %s.", volumeName, DefaultTimeUntilDeletion.String())
```

Comment:
The garbage collection here is for all attachments of an AzVolume and not the AzVolume itself. The log message is confusing.